templates/flutter/lib/src/realtime_mixin.dart.twig

```diff
@@ -118,43 +122,49 @@ mixin RealtimeMixin {
       port: uri.port,
       queryParameters: {
         "project": client.config['project'],
-        "channels[]": _channels.keys.toList(),
+        "channels[]": _channels.toList(),
       },
       path: uri.path + "/realtime",
     );
   }
 
   RealtimeSubscription subscribeTo(List<String> channels) {
     StreamController<RealtimeMessage> controller = StreamController.broadcast();
-    for(var channel in channels) {
-      if (!_channels.containsKey(channel)) {
-        _channels[channel] = [];
-      }
-      _channels[channel]!.add(controller);
-    }
+    _channels.addAll(channels);
     Future.delayed(Duration.zero, () => _createSocket());
+    int counter = _subscriptionsCounter++;
     RealtimeSubscription subscription = RealtimeSubscription(
-        stream: controller.stream,
+        controller: controller,
+        channels: channels,
         close: () async {
+          _subscriptions.remove(counter);
+          _subscriptionsCounter--;
           controller.close();
-          for(var channel in channels) {
-            _channels[channel]!.remove(controller);
-            if (_channels[channel]!.isEmpty) {
-              _channels.remove(channel);
-            }
-          }
-          if(_channels.isNotEmpty) {
+          _cleanup(channels);
+
+          if (_channels.isNotEmpty) {
             await Future.delayed(Duration.zero, () => _createSocket());
           } else {
             await _closeConnection();
           }
         });
+    _subscriptions[counter] = subscription;
```

Comment:
I think using a counter can lead to problems. For example:

0. start (counter = 0, subscriptions map is empty)
1. subscribe() (counter = 1, subscriptions[0] is set)
2. subscribe() (counter = 2, subscriptions[0] and [1] are set)
3. the first one gets removed (counter = 1, subscriptions[1] is set)
4. a new subscription added (counter = 2, subscriptions[1] is set and replaces the subscription from step 2)

Reply:
Thanks, I've updated to use the time in microseconds to overcome this. please let me know what you think.